pkg/cloud/services/networking/securitygroups.go

```diff
@@ -399,6 +402,67 @@ func (s *Service) createRule(r infrav1.SecurityGroupRule) (infrav1.SecurityGroup
 	return convertOSSecGroupRuleToConfigSecGroupRule(*rule), nil
 }
 
+func contains(list []string, element string) bool {
+	for _, e := range list {
+		if e == element {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *Service) ReconcileSecurityGroupsToInstanceAttachedInterfaces(logger logr.Logger, openStackMachine *infrav1.OpenStackMachine, attachedInterfaces []attachinterfaces.Interface, desiredSecurityGroupIDs []string) error {
+	for _, iface := range attachedInterfaces {
+		portUpdateNeeded := false
+		portID := iface.PortID
+		port, err := s.client.GetPort(portID)
+		if err != nil {
+			return fmt.Errorf("error getting port %s: %v", portID, err)
+		}
+
+		// check if port already has desired security groups
+		for _, securityGroupID := range desiredSecurityGroupIDs {
+			if !contains(port.SecurityGroups, securityGroupID) {
+				logger.Info("Port doesn't have desired security group", "portID", portID, "securityGroupID", securityGroupID)
+				portUpdateNeeded = true
+				break
+			}
+		}
+		// check if port has security groups that are not desired
+		for _, securityGroupID := range port.SecurityGroups {
+			if !contains(desiredSecurityGroupIDs, securityGroupID) {
+				logger.Info("Port has security group that is not desired", "portID", portID, "securityGroupID", securityGroupID)
+				portUpdateNeeded = true
+				break
```

Comment:
Hmm. Is it a bug that the default security group is attached to the ports?

Reply:
Ports that are created when port security is enabled will have the `default` security group itself managed by Neutron.

I see two options:
* We keep it and look for its ID (once and early to avoid many Neutron API calls) and put it in `desiredSecuritygroupIDs`, therefore it's not a problem anymore.
* We remove it in this loop

I personally prefer option 1 for safety & because that's default OpenStack behaviours here.

cc @stephenfin with your OpenStack dev hat, any preference?

Reply:
tbh, until today I didn't know that adding a port would add the default security group. What is the source of these ports? If they are ports that we are creating ourselves then we can instruct neutron not to assign a security group to the port. From openstackclient this is done by e.g.:

```
openstack port create --no-security-group ...
```

which under the hood sets the `security_group_ids` field to `[]` (empty list) when creating the port. Could we do the same thing?

If these ports are coming from elsewhere then it's possible to have other security groups than the default one assigned to the port (`openstack port create --security-group <sec-group> ...`) so I'm not sure how we can ignore everything? Do we forbid specifying ports with security groups defined?

Reply:
Given the above, we probably need to decide what's authoritative: the security groups on the port, or the security ports defined on the `Machine`. I assume the latter, but if this is the case then we're going to modify something that the user gave us. If we go with the former, we're either modifying or ignoring something *else* that the user gave us.

Reply:
> What is the source of these ports?

This is where we pre-build ports before creating an instance:
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/7961f21443918d50f2567a724b5ceeec4b445428/pkg/cloud/services/compute/instance.go#L177-L222

We don't pass any security groups.

> Could we do the same thing?

I don't think so because we don't provide a list but a list of `SecurityGroupFilter`, so I suspect that nothing will be passed to Neutron API if we pass an empty list of filters.

> we probably need to decide what's authoritative:

I want CAPO to have the authority if possible: if users provide a security group for the Machine, then we add it to the port in the reconcile (not at instance creation) - so instance + initial ports have no SGs in theory.

Reply:
I checked with @pierreprinetti and the good news is that `port.CreateOpts.SecurityGroups` in Gophercloud is `*[]string` so if we pass an empty slice of strings, the port should be created with no security group. I'll test it.